include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,401 @@
+                    const size_t maxGenerations,
+                    const double crossoverProb,
+                    const double mutationProb,
+                    const double mutationStrength,
+                    const double epsilon,
+                    const arma::vec lowerBound,
+                    const arma::vec upperBound) :
+    populationSize(populationSize),
+    maxGenerations(maxGenerations),
+    crossoverProb(crossoverProb),
+    mutationProb(mutationProb),
+    mutationStrength(mutationStrength),
+    epsilon(epsilon),
+    lowerBound(lowerBound),
+    upperBound(upperBound)
+{ /* Nothing to do here. */ }
+
+//! Optimize the function.
+template<typename MatType,
+         typename... ArbitraryFunctionType,
+         typename... CallbackTypes>
+std::vector<MatType> NSGA2::Optimize(std::tuple<ArbitraryFunctionType...>& objectives,
+                                     MatType& iterate,
+                                     CallbackTypes&&... callbacks)
+{
+  // Make sure for evolution to work at least four candidates are present.
+  if (populationSize < 4 && populationSize % 4 != 0)
+  {
+    throw std::logic_error("NSGA2::Optimize(): population size should be at"
+        " least 4, and, a multiple of 4!");
+  }
+
+  numObjectives = sizeof...(ArbitraryFunctionType);
+
+  // Cache calculated objectives.
+  std::vector<arma::Col<double> > calculatedObjectives;
+  // Pre-allocate space for the calculated objectives.
+  calculatedObjectives.resize(populationSize);
+
+  // Population size reserved to 2 * populationSize + 1 to accommodate
+  // for the size  of intermediate candidate population.
+  std::vector<MatType> population;
+  population.reserve(2 * populationSize + 1);
+
+  // Pareto fronts, initialized during non-dominated sorting.
+  std::vector<std::vector<size_t> > fronts;
+  // Initialised in CrowdingDistanceAssignment.
+  std::vector<double> crowdingDistance;
+  // Initialised during non-dominated sorting.
+  std::vector<size_t> ranks;
+
+  // Controls early termination of the optimization process.
+  bool terminate = false;
+
+  // Generate the population based on a Gaussian distribution around the given
+  // starting point.
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    population.push_back(arma::randu<MatType>(iterate.n_rows,
+        iterate.n_cols) + iterate);
+  }
+
+  Info << "NSGA2 initialized successfully. Optimization started." << std::endl;
+
+  // Evaluate the fitness before optimization.
+  for (size_t i = 0; i < population.size(); i++)
+    calculatedObjectives[i] = arma::Col<double>(numObjectives);
+  EvaluateObjectives(population, objectives, calculatedObjectives);
+
+  // Iterate until maximum number of generations is obtained.
+  terminate |= Callback::BeginOptimization(*this, objectives, iterate, callbacks...);
+
+  for (size_t generation = 1; generation <= maxGenerations && !terminate; generation++)
+  {
+    Info << "Generation number: " << generation << std::endl;
```

Comment:
@zoq the `calculatedObjectives` holds the objectives for the entire population, will I print that?
(Since it'll have entries for each of the objectives for each candidate in the population.)